File: vw_c/src/Vulkan/Instance.cpp
#include "Vulkan/Instance.h"

#include <VulkanWrapper/Vulkan/DeviceFinder.h>
#include <VulkanWrapper/Vulkan/Instance.h>

vw::Instance *vw_create_instance(const VwInstanceCreateArguments *arguments) {
    std::span extensions_as_span(arguments->extensions,
                                 arguments->extensions_count);

    if (arguments->debug_mode)
        return new vw::Instance{vw::InstanceBuilder()
                                    .addExtensions(extensions_as_span)
                                    .addPortability()
                                    .setDebug()
                                    .build()};
    return new vw::Instance{vw::InstanceBuilder()
                                .addExtensions(extensions_as_span)
                                .addPortability()
                                .build()};
}

vw::DeviceFinder *vw_find_gpu_from_instance(const vw::Instance *instance) {
    return new vw::DeviceFinder{instance->findGpu()};
}

void vw_destroy_instance(vw::Instance *instance) { delete instance; }

void vw_destroy_device_finder(vw::DeviceFinder *deviceFinder) {
    delete deviceFinder;
}

File: vw_py/Vulkan/Instance.py
import bindings_vw_py as bindings
from Vulkan import DeviceFinder

class Instance:
    def __init__(self, instance):
        self.instance = instance

    def find_gpu(self):
        return DeviceFinder.DeviceFinder(self, bindings.vw_find_gpu_from_instance(self.instance))

    def __del__(self):
        print("remove instance")
        bindings.vw_destroy_instance(self.instance)

class InstanceBuilder:
    extensions = bindings.CStringArray()
    debug_mode = False

    def __init__(self):
        self.extensions = bindings.CStringArray()
        self.debug_mode = True

    def add_extensions(self, extensions):
        self.extensions.push(extensions)
        return self

    def set_debug_mode(self):
        debug_mode = True
        return self

    def build(self):
        arguments = bindings.VwInstanceCreateArguments()
        arguments.debug_mode = self.debug_mode
        arguments.extensions = bindings.VwStringArray(self.extensions)
        instance = bindings.vw_create_instance(arguments)
        return Instance(instance)
